shareclasses:enableBCI leads to ClassNotFoundException with 11.0.7

Java -version output
--------------------

Output from `java -version`.
If the shas are not public, include equivalent public shas.

```
openjdk version "11.0.7" 2020-04-14
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.7+10)
Eclipse OpenJ9 VM AdoptOpenJDK (build openj9-0.20.0, JRE 11 Linux amd64-64-Bit Compressed References 20200416_574 (JIT enabled, AOT enabled)
OpenJ9 - 05fa2d361
OMR - d4365f371
JCL - 838028fc9d based on jdk-11.0.7+10)
```

Summary of problem
------------------

The application is running on adoptopenjdk/openjdk11-openj9:alpine-jre base image.

We provide a javaagent doing byte code instrumentation. Our client is running the above mentioned JDK version and the following shared class cache settings:

```
-Xshareclasses:name=<name>
-XX:SharedCacheHardLimit=300m
-Xshareclasses:expire=240
-Xshareclasses:enableBCI
-Xshareclasses:persistent
-Xshareclasses:nonfatal
-Xshareclasses:groupAccess
```

However the the javaagent fails to start with the following exception:

```
Could not start Instana agent: java.com.instana.agent.log.v1.LogDispatcher
java.lang.NoClassDefFoundError: java.com.instana.agent.log.v1.LogDispatcher
at com.instana.agent.loader.v1_3_26.AgentLoader.setupAgent(AgentLoader.java:137)
at com.instana.agent.loader.v1_3_26.AgentLoader.agentmain(AgentLoader.java:73)
at com.instana.agent.loader.v1_3_26.AgentLoader.agentmain(AgentLoader.java:65)
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
at java.base/java.lang.reflect.Method.invoke(Unknown Source)
at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndStartAgent(Unknown Source)
at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndCallAgentmain(Unknown Source)
at java.base/openj9.internal.tools.attach.target.Attachment.loadAgentLibraryImpl(Native Method)
at java.base/openj9.internal.tools.attach.target.Attachment.loadAgentLibrary(Unknown Source)
at java.base/openj9.internal.tools.attach.target.Attachment.parseLoadAgent(Unknown Source)
at java.base/openj9.internal.tools.attach.target.Attachment.doCommand(Unknown Source)
at java.base/openj9.internal.tools.attach.target.Attachment.run(Unknown Source)
Caused by: java.lang.ClassNotFoundException: java.com.instana.agent.log.v1.LogDispatcher
at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(Unknown Source)
at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(Unknown Source)
at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
... 14 more
Exception in thread "Attachment 41301" java.lang.reflect.InvocationTargetException
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
at java.base/java.lang.reflect.Method.invoke(Unknown Source)
at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndStartAgent(Unknown Source)
at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndCallAgentmain(Unknown Source)
at java.base/openj9.internal.tools.attach.target.Attachment.loadAgentLibraryImpl(Native Method)
at java.base/openj9.internal.tools.attach.target.Attachment.loadAgentLibrary(Unknown Source)
at java.base/openj9.internal.tools.attach.target.Attachment.parseLoadAgent(Unknown Source)
at java.base/openj9.internal.tools.attach.target.Attachment.doCommand(Unknown Source)
at java.base/openj9.internal.tools.attach.target.Attachment.run(Unknown Source)
Caused by: java.lang.RuntimeException: java.lang.NoClassDefFoundError: java.com.instana.agent.log.v1.LogDispatcher
at com.instana.agent.loader.v1_3_26.AgentLoader.setupAgent(AgentLoader.java:168)
at com.instana.agent.loader.v1_3_26.AgentLoader.agentmain(AgentLoader.java:73)
at com.instana.agent.loader.v1_3_26.AgentLoader.agentmain(AgentLoader.java:65)
... 11 more
Caused by: java.lang.NoClassDefFoundError: java.com.instana.agent.log.v1.LogDispatcher
at com.instana.agent.loader.v1_3_26.AgentLoader.setupAgent(AgentLoader.java:137)
... 13 more
Caused by: java.lang.ClassNotFoundException: java.com.instana.agent.log.v1.LogDispatcher
at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(Unknown Source)
at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(Unknown Source)
at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
... 14 more
Agent failed to start!
JVMJ9TI064E Agent initialization function Agent_OnAttach failed for library instrument, return code 102
```
with the class `java.com.instana.agent.log.v1.LogDispatcher` injected into the boot classloader during javaagent initialisation.

This used to work for older versions of openJ9 and as a workaround this also works with `-Xshareclasses:none` on the above mentioned JDK version.


Diagnostic files
----------------
<none>

--- a/runtime/bcutil/ROMClassBuilder.cpp
+++ b/runtime/bcutil/ROMClassBuilder.cpp
@@ -543,12 +543,18 @@ ROMClassBuilder::prepareAndLaydown( BufferManager *bufferManager, ClassFileParse
 #if defined(J9VM_OPT_SHARED_CLASSES)
 	if (context->isROMClassShareable()) {
 		UDATA loadType = J9SHR_LOADTYPE_NORMAL;
-		if (context->isClassHidden()) {
-			loadType = J9SHR_LOADTYPE_NOT_FROM_PATH;
-		} else if (context->isRedefining()) {
+		if (context->isRedefining()) {
 			loadType = J9SHR_LOADTYPE_REDEFINED;
 		} else if (context->isRetransforming()) {
 			loadType = J9SHR_LOADTYPE_RETRANSFORMED;
+		} else if (context->isClassUnsafe()
+			|| context->isClassHidden()
+			|| (LOAD_LOCATION_UNKNOWN == context->loadLocation())
+		) {
+			/* For redefining/transforming, we still want loadType to be J9SHR_LOADTYPE_REDEFINED/J9SHR_LOADTYPE_RETRANSFORMED,
+			 * so put these checks after the redefining/transforming checks.
+			 */ 
+			loadType = J9SHR_LOADTYPE_NOT_FROM_PATH;
 		}
 
 		SCStoreTransaction sharedStoreClassTransaction =

--- a/runtime/shared_common/SCImplementedAPI.cpp
+++ b/runtime/shared_common/SCImplementedAPI.cpp
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2001, 2020 IBM Corp. and others
+ * Copyright (c) 2001, 2021 IBM Corp. and others
  *
  * This program and the accompanying materials are made available under
  * the terms of the Eclipse Public License 2.0 which accompanies this
@@ -478,7 +478,7 @@ j9shr_classStoreTransaction_start(void * tobj, J9VMThread* currentThread, J9Clas
 	modContext = sconfig->modContext;
 
 	if ((classloader != NULL)
-		&& (J9SHR_LOADTYPE_NOT_FROM_PATH != loadType) /* no need to set classpath if loadType is J9SHR_LOADTYPE_NOT_FROM_PATH */
+		&& (J9SHR_LOADTYPE_NORMAL == loadType) /* no need to set classpath if loadType is not J9SHR_LOADTYPE_NORMAL */
 	) {
 		/* default values for bootstrap: */
 		IDATA helperID = 0;

